README.md

```diff
@@ -30,153 +31,238 @@ addOne(2); // 3
+## The `?` Placeholder Argument
+
+The `?` placeholder token can be supplied one or more times at the top level of the argument list of
+a _CallExpression_, _CallMemberExpression_, or `new` (e.g. `f~(?)` or `o.f~(?)`). `?` is **not**
+an expression, rather it is a syntactic element that indicates special behavior (much like how
+`` `...` AssignmentExpression `` indicates spread, yet is itself not an expression).
 
 ```js
 // valid
-f(x, ?)           // partial application from left
-f(?, x)           // partial application from right
-f(?, x, ?)        // partial application for any arg
-o.f(x, ?)         // partial application from left
-o.f(?, x)         // partial application from right
-o.f(?, x, ?)      // partial application for any arg
-super.f(?)        // partial application allowed for call on |SuperProperty|
+f~(x, ?)          // partial application from left
+f~(?, x)          // partial application from right
+f~(?, x, ?)       // partial application for any arg
+o.f~(x, ?)        // partial application from left
+o.f~(?, x)        // partial application from right
+o.f~(?, x, ?)     // partial application for any arg
+super.f~(?)       // partial application allowed for call on |SuperProperty|
+new C~(?)         // partial application of constructor
 
 // invalid
-f(x + ?)          // `?` not in top-level Arguments of call
+f~(x + ?)         // `?` not in top-level Arguments of call
 x + ?             // `?` not in top-level Arguments of call
-?.f()             // `?` not in top-level Arguments of call
-new f(?)          // `?` not supported in `new`
-super(?)          // `?` not supported in |SuperCall|
+?.f~()            // `?` not in top-level Arguments of call
+super~(?)         // `?` not supported in |SuperCall|
 ```
 
-# Semantics
+In addition the `?` placeholder token can be followed by a decimal integer value &ge; 0 indicating a fixed
+ordinal position (i.e., `?0`). Ordinal placeholders can be repeated multiple times within a partial application,
+but still point to the same unbound argument. If a partial application contains a mix of non-ordinal placeholders 
+and ordinal placeholders, the non-ordinal placeholders are implicitly numbered to fill in any gaps:
 
-The `?` placeholder token can only be used in an argument list of a call expression, or as the only
-token in a placeholder of a template expression or tagged template expression. When present, the 
-result of the call is a new function with a parameter for each `?` token in the argument list. 
-Any non-placeholder expression in the argument list becomes fixed in its position. This is 
-illustrated by the following syntactic conversion:
+```js
+const printArgs = (a = "a", b = "b", c = "c") => console.log(`${a}, ${b}, ${c}`);
+printArgs();                            // prints: a, b, c
+printArgs(1, 2, 3);                     // prints: 1, 2, 3
+
+const swapAC = printArgs~(?2, ?, ?0);
```

Comment:
For your first comment example, do you mean:

```diff
-// g is now (roughly): (_2, _0, _1) => f(_2, _0, 1, _1);
+// g is now (roughly): (_0, _1, _2) => f(_2, _0, 1, _1);
```

Reply:
That's a fair point, and my current explanation is *also* a refactoring hazard. In the explainer currently, any ordinal placeholder reserves a position that a normal placeholder would otherwise occupy:

```js
f~(?, ?, ?) // implicitly f~(?0, ?1, ?2)
f~(?0, ?, ?) // implicitly f~(?0, ?1, ?2)
f~(?, ?0, ?) // implicitly f~(?1, ?0, ?2)
```

Its possibly better for normal placeholders to always just increment from 0, as this works better with refactoring:

```js
// above examples
f~(?, ?, ?) // implicitly f~(?0, ?1, ?2)
f~(?0, ?, ?) // implicitly f~(?0, ?0, ?1)
f~(?, ?0, ?) // implicitly f~(?0, ?0, ?1)

// refactoring f~(?, ?, ?) by inserting an ordinal argument anywhere
// `^` - inserted
// `=` - same

   f~(?0, ?, ?, ?)     // f~(?0, ?0, ?1, ?2)
//    ^^  =======            ^^  ==========

   f~(?, ?1, ?, ?)     // f~(?0, ?1, ?1, ?2)
//    =  ^^  ====            ==  ^^  ======

   f~(?, ?, ?5, ?)     // f~(?0, ?1, ?5, ?2)
//    ====  ^^  =            ======  ^^  ==
```

We *could* ban mixing them, but as long as we choose a consistent numbering mechanism there's no reason to ban mixing them outright in the language. A linter could ban mixing them instead, leaving it open to override a lint rule in special circumstances.